File: Battleships/HumanPlayer.cpp
#include "HumanPlayer.hpp"
#include "Color.hpp"

namespace {
    const int shipSizes[] = { 4, 3, 3, 2, 2, 2, 1, 1, 1, 1 };
    const int BOARD_SIZE = 10;
    const int MAX_SINGLE_PLACEMENT_TRIES = 100;
    const int AROUND_OFFSET = 1;
}

void HumanPlayer::placeShips() {
    std::cout << "Расстановка кораблей для игрока: " << this->name << std::endl;
    int choice;
    while (true) {
        std::cout << "Выберите способ расстановки:\n";
        Color::setColor(Color::GREEN);
        std::cout << "1 - Автоматическая расстановка\n";
        Color::setColor(Color::YELLOW);
        std::cout << "2 - Ручная расстановка\n";
        Color::resetColor();
        std::cout << "Ваш выбор: ";
        std::cin >> choice;
        if (choice == 1) {
            if (automaticPlacement()) {
                break;
            }
            else {
                Color::setColor(Color::RED);
                std::cout << "Автоматическая расстановка не удалась. Попробуйте ручную расстановку.\n";
                Color::resetColor();
            }
        }
        else if (choice == 2) {
            manualPlacement();
            break;
        }
        else {
            Color::setColor(Color::RED);
            std::cout << "Неверный выбор! Попробуйте снова.\n";
            Color::resetColor();
        }
    }
}

void HumanPlayer::makeMove(Player& enemy) {
    makeMoveWithResult(enemy);
}


bool HumanPlayer::makeMoveWithResult(Player& enemy) {
    std::cout << "\n=== Ход игрока. " << this->name << " ===" << std::endl;

    Color::setColor(Color::GREEN);
    std::cout << "Ваше поле:" << std::endl;
    Color::resetColor();
    this->myBoard.display(true);

    Color::setColor(Color::BLUE);
    std::cout << "\nПоле противника:" << std::endl;
    Color::resetColor();
    this->enemyBoard.display(false);

    Coordinate target = inputCoordinate();
    ShotResult result = enemy.getShotResult(target);
    bool wasHit = false;

    switch (result) {
    case ShotResult::Miss:
        this->enemyBoard.setCellState(target, CellState::Miss);
        Color::setColor(Color::BLUE);
        std::cout << "Промах!" << std::endl;
        Color::resetColor();
        wasHit = false;
        break;
    case ShotResult::Hit:
        this->enemyBoard.setCellState(target, CellState::Hit);
        Color::setColor(Color::YELLOW);
        std::cout << "Попадание!" << std::endl;
        Color::resetColor();
        wasHit = true;
        break;
    case ShotResult::Sunk:
        this->enemyBoard.setCellState(target, CellState::Hit);
        Color::setColor(Color::RED);
        std::cout << "Уничтожен корабль!" << std::endl;
        Color::resetColor();
        wasHit = true;
        markAreaAroundDestroyedShip(enemy, target);
        break;
    }
    return wasHit;
}

void HumanPlayer::markSurroundingCells(const Coordinate& center) {
    for (int dy = -AROUND_OFFSET; dy <= AROUND_OFFSET; ++dy) {
        for (int dx = -AROUND_OFFSET; dx <= AROUND_OFFSET; ++dx) {
            Coordinate around(center.x + dx, center.y + dy);
            if (around.x >= 0 && around.x < BOARD_SIZE && around.y >= 0 && around.y < BOARD_SIZE) {
                if (this->enemyBoard.getCellState(around) == CellState::Empty) {
                    this->enemyBoard.setCellState(around, CellState::Miss);
                }
            }
        }
    }
}

void HumanPlayer::markAreaAroundDestroyedShip(Player& enemy, const Coordinate& hitCoord) {
    markSurroundingCells(hitCoord);
    for (int radius = 1; radius <= 3; radius++) {
        for (int dy = -radius; dy <= radius; ++dy) {
            for (int dx = -radius; dx <= radius; ++dx) {
                Coordinate around(hitCoord.x + dx, hitCoord.y + dy);
                if (around.x >= 0 && around.x < BOARD_SIZE && around.y >= 0 && around.y < BOARD_SIZE) {
                    if (this->enemyBoard.getCellState(around) == CellState::Hit) {
                        markSurroundingCells(around);
                    }
                }
            }
        }
    }
}

Coordinate HumanPlayer::inputCoordinate() const {
    Coordinate coord;
    std::string input;

    while (true) {
        try {
            std::cout << "Введите координаты для выстрела (например, A1): ";
            input = getSafeInput();

            if (parseCoordinate(input, coord)) {

                CellState state = this->enemyBoard.getCellState(coord);
                if (state == CellState::Empty) {
                    return coord;
                }
                else {
                    Color::setColor(Color::RED);
                    std::cout << "Вы уже стреляли в эту клетку!" << std::endl;
                    Color::resetColor();
                }
            }
            
        }
        catch (const std::exception& e) {
            Color::setColor(Color::RED);
            std::cout << "Ошибка ввода: " << e.what() << std::endl;
            Color::resetColor();
        }
    }
}


void HumanPlayer::manualPlacement() {
    const char* shipNames[] = { "четырехпалубный", "трехпалубный", "трехпалубный",
                              "двухпалубный", "двухпалубный", "двухпалубный",
                              "однопалубный", "однопалубный", "однопалубный", "однопалубный" };

    for (int i = 0; i < BOARD_SIZE; ++i) {
        int size = shipSizes[i];
        bool placed = false;

        while (!placed) {
            try {
                Color::setColor(Color::YELLOW);
                std::cout << "\nРазмещение " << shipNames[i] << " корабля (" << size << " палубы)" << std::endl;
                Color::resetColor();

                myBoard.display(true);

                Coordinate startCoord;
                bool validStart = false;

                while (!validStart) {
                    std::cout << "Введите начальную координату (например, A1): ";
                    std::string input = getSafeInput();

                    if (parseCoordinate(input, startCoord)) {
                        validStart = true;
                    }
                    else {
                        Color::setColor(Color::RED);
                        std::cout << "Неверный формат координат! Используйте формат A1-J10." << std::endl;
                        Color::resetColor();
                    }
                }

                if (size > 1) {
                    char orient;
                    bool validOrientation = false;

                    while (!validOrientation) {
                        std::cout << "Выберите ориентацию (H - горизонтально, V - вертикально): ";
                        std::string orientInput = getSafeInput();

                        if (!orientInput.empty()) {
                            orient = toupper(orientInput[0]);
                            if (orient == 'H' || orient == 'V') {
                                validOrientation = true;
                            }
                            else {
                                Color::setColor(Color::RED);
                                std::cout << "Введите H или V!" << std::endl;
                                Color::resetColor();
                            }
                        }
                    }

                    Orientation orientation = (orient == 'H') ? Orientation::Horizontal : Orientation::Vertical;
                    Ship ship(size, startCoord, orientation);

                    if (isValidShipPlacement(size, startCoord, orientation)) {
                        if (myBoard.placeShip(ship)) {
                            ships.push_back(ship);
                            placed = true;
                            Color::setColor(Color::GREEN);
                            std::cout << "Корабль размещен успешно!" << std::endl;
                            Color::resetColor();
                        }
                        else {
                            Color::setColor(Color::RED);
                            std::cout << "Ошибка при размещении корабля!" << std::endl;
                            Color::resetColor();
                        }
                    }
                    else {
                        Color::setColor(Color::RED);
                        std::cout << "Неверное размещение корабля! Попробуйте снова." << std::endl;
                        Color::resetColor();
                    }
                }
                else {
                    Ship ship(size, startCoord, Orientation::Horizontal);

                    if (isValidShipPlacement(size, startCoord, Orientation::Horizontal)) {
                        if (myBoard.placeShip(ship)) {
                            ships.push_back(ship);
                            placed = true;
                            Color::setColor(Color::GREEN);
                            std::cout << "Корабль размещен успешно!" << std::endl;
                            Color::resetColor();
                        }
                    }
                    else {
                        Color::setColor(Color::RED);
                        std::cout << "Неверное размещение корабля! Попробуйте снова." << std::endl;
                        Color::resetColor();
                    }
                }
            }
            catch (const std::exception& e) {
                Color::setColor(Color::RED);
                std::cout << "Ошибка: " << e.what() << std::endl;
                Color::resetColor();
            }
        }
    }
}

bool HumanPlayer::automaticPlacement() {
    int attempts = 0;
    const int MAX_ATTEMPTS = 5;

    while (attempts < MAX_ATTEMPTS) {
        myBoard.clearBoard();
        ships.clear();

        bool success = true;

        for (int size : shipSizes) {
            bool placed = false;
            int placementAttempts = 0;

            while (!placed && placementAttempts < MAX_SINGLE_PLACEMENT_TRIES) {
                int x = std::rand() % BOARD_SIZE;
                int y = std::rand() % BOARD_SIZE;
                Orientation orientation = (std::rand() % 2 == 0) ? Orientation::Horizontal : Orientation::Vertical;

                Ship ship(size, Coordinate(x, y), orientation);
                if (myBoard.isValidPlacement(ship)) {
                    if (myBoard.placeShip(ship)) {
                        ships.push_back(ship);
                        placed = true;
                    }
                }
                placementAttempts++;
            }

            if (!placed) {
                success = false;
                break;
            }
        }

        if (success) {
            Color::setColor(Color::GREEN);
            std::cout << "Автоматическая расстановка завершена!" << std::endl;
            Color::resetColor();
            myBoard.display(true);

            if (attempts > 0) {
                std::cout << "Потребовалось " << (attempts + 1) << " попыток для успешной расстановки." << std::endl;
            }

            return true;
        }

        attempts++;
    }

    Color::setColor(Color::RED);
    std::cout << "Не удалось автоматически расставить корабли после " << MAX_ATTEMPTS << " попыток." << std::endl;
    Color::resetColor();
    return false;
}

bool HumanPlayer::parseCoordinate(const std::string& input, Coordinate& coord) {
    coord = Coordinate(-1, -1);

    if (input.empty()) {
        return false;
    }

    std::string str = input;

    size_t start = str.find_first_not_of(" \t");
    if (start == std::string::npos) {
        return false;
    }
    size_t end = str.find_last_not_of(" \t");
    str = str.substr(start, end - start + 1);

    if (str.length() < 2) {
        return false;
    }

    std::transform(str.begin(), str.end(), str.begin(), ::toupper);

    std::string cleaned;
    for (char c : str) {
        if (c != ' ' && c != '-' && c != ':') {
            cleaned += c;
        }
    }

    if (cleaned.empty()) {
        return false;
    }

    char letter = cleaned[0];
    if (letter < 'A' || letter > 'J') {
        return false;
    }

    std::string numberStr = cleaned.substr(1);
    if (numberStr.empty()) {
        return false;
    }

    for (char c : numberStr) {
        if (!isdigit(c)) {
            return false;
        }
    }

    int number;
    try {
        number = std::stoi(numberStr);
    }
    catch (const std::exception&) {
        return false;
    }

    if (number < 1 || number > 10) {
        return false;
    }

    coord.x = letter - 'A';
    coord.y = number - 1;

    return true;
}

std::string HumanPlayer::getSafeInput() {
    std::string input;
    std::getline(std::cin, input);
    if (std::cin.fail()) {
        std::cin.clear();
        std::cin.ignore((std::numeric_limits<std::streamsize>::max)(), '\n');
    }
    return input;
}

bool HumanPlayer::isValidShipPlacement(int size, const Coordinate& start, Orientation orientation) const {
    Ship tempShip(size, start, orientation);
    return myBoard.isValidPlacement(tempShip);
}

File: SeaBattleCSharp/HumanPlayer.cs
using System;
using System.Linq;

namespace SeaBattleCSharp
{
    public class HumanPlayer : Player
    {
        private readonly int[] shipSizes = { 4, 3, 3, 2, 2, 2, 1, 1, 1, 1 };
        private const int BOARD_SIZE = 10;
        private const int MAX_SINGLE_PLACEMENT_TRIES = 100;

        public HumanPlayer(string name) : base(name) { }

        public override bool IsHuman() => true;

        public override void PlaceShips()
        {
            Console.WriteLine($"Расстановка кораблей для игрока: {name}");

            int choice;
            while (true)
            {
                Console.WriteLine("Выберите способ расстановки:");
                Color.SetColor(Color.GREEN);
                Console.WriteLine("1 - Автоматическая расстановка");
                Color.SetColor(Color.YELLOW);
                Console.WriteLine("2 - Ручная расстановка");
                Color.ResetColor();
                Console.Write("Ваш выбор: ");

                if (int.TryParse(Console.ReadLine(), out choice))
                {
                    if (choice == 1)
                    {
                        if (AutomaticPlacement())
                            break;
                        else
                        {
                            Color.SetColor(Color.RED);
                            Console.WriteLine("Автоматическая расстановка не удалась. Попробуйте ручную расстановку.");
                            Color.ResetColor();
                        }
                    }
                    else if (choice == 2)
                    {
                        ManualPlacement();
                        break;
                    }
                    else
                    {
                        Color.SetColor(Color.RED);
                        Console.WriteLine("Неверный выбор! Попробуйте снова.");
                        Color.ResetColor();
                    }
                }
            }
        }

        public override void MakeMove(Player enemy)
        {
            MakeMoveWithResult(enemy);
        }

        public override bool MakeMoveWithResult(Player enemy)
        {
            Console.WriteLine($"\n=== Ход игрока. {name} ===");

            Color.SetColor(Color.GREEN);
            Console.WriteLine("Ваше поле:");
            Color.ResetColor();
            myBoard.Display(true);

            Color.SetColor(Color.BLUE);
            Console.WriteLine("\nПоле противника:");
            Color.ResetColor();
            enemyBoard.Display(false);

            Coordinate target = InputCoordinate();
            ShotResult result = enemy.GetShotResult(target);
            bool wasHit = false;

            switch (result)
            {
                case ShotResult.Miss:
                    enemyBoard.SetCellState(target, CellState.Miss);
                    Color.SetColor(Color.BLUE);
                    Console.WriteLine("Промах!");
                    Color.ResetColor();
                    wasHit = false;
                    break;

                case ShotResult.Hit:
                    enemyBoard.SetCellState(target, CellState.Hit);
                    Color.SetColor(Color.YELLOW);
                    Console.WriteLine("Попадание!");
                    Color.ResetColor();
                    wasHit = true;
                    break;

                case ShotResult.Sunk:
                    enemyBoard.SetCellState(target, CellState.Hit);
                    Color.SetColor(Color.RED);
                    Console.WriteLine("Уничтожен корабль!");
                    Color.ResetColor();
                    wasHit = true;
                    MarkAreaAroundDestroyedShip(enemy, target);
                    break;
            }

            return wasHit;
        }

        public override void MarkAreaAroundDestroyedShip(Player enemy, Coordinate hitCoord)
        {
            MarkSurroundingCells(hitCoord);

            for (int radius = 1; radius <= 3; radius++)
            {
                for (int dy = -radius; dy <= radius; dy++)
                {
                    for (int dx = -radius; dx <= radius; dx++)
                    {
                        Coordinate around = new Coordinate(hitCoord.X + dx, hitCoord.Y + dy);
                        if (around.X >= 0 && around.X < BOARD_SIZE &&
                            around.Y >= 0 && around.Y < BOARD_SIZE)
                        {
                            if (enemyBoard.GetCellState(around) == CellState.Hit)
                            {
                                MarkSurroundingCells(around);
                            }
                        }
                    }
                }
            }
        }

        private void MarkSurroundingCells(Coordinate center)
        {
            for (int dy = -1; dy <= 1; dy++)
            {
                for (int dx = -1; dx <= 1; dx++)
                {
                    Coordinate around = new Coordinate(center.X + dx, center.Y + dy);
                    if (around.X >= 0 && around.X < BOARD_SIZE &&
                        around.Y >= 0 && around.Y < BOARD_SIZE)
                    {
                        if (enemyBoard.GetCellState(around) == CellState.Empty)
                        {
                            enemyBoard.SetCellState(around, CellState.Miss);
                        }
                    }
                }
            }
        }

        private Coordinate InputCoordinate()
        {
            while (true)
            {
                Console.Write("Введите координаты для выстрела (например, A1): ");
                string input = Console.ReadLine()?.ToUpper();

                if (string.IsNullOrEmpty(input) || input.Length < 2)
                {
                    Color.SetColor(Color.RED);
                    Console.WriteLine("Неверный формат!");
                    Color.ResetColor();
                    continue;
                }

                char letter = input[0];
                if (!int.TryParse(input.Substring(1), out int number))
                {
                    Color.SetColor(Color.RED);
                    Console.WriteLine("Неверный формат числа!");
                    Color.ResetColor();
                    continue;
                }

                int x = letter - 'A';
                int y = number - 1;

                if (x >= 0 && x < BOARD_SIZE && y >= 0 && y < BOARD_SIZE)
                {
                    Coordinate coord = new Coordinate(x, y);
                    CellState state = enemyBoard.GetCellState(coord);

                    if (state == CellState.Empty)
                        return coord;
                    else
                    {
                        Color.SetColor(Color.RED);
                        Console.WriteLine("Вы уже стреляли в эту клетку!");
                        Color.ResetColor();
                    }
                }
                else
                {
                    Color.SetColor(Color.RED);
                    Console.WriteLine("Неверные координаты! Используйте формат A1-J10.");
                    Color.ResetColor();
                }
            }
        }

        private void ManualPlacement()
        {
            string[] shipNames = { "четырехпалубный", "трехпалубный", "трехпалубный",
                                   "двухпалубный", "двухпалубный", "двухпалубный",
                                   "однопалубный", "однопалубный", "однопалубный", "однопалубный" };

            for (int i = 0; i < shipSizes.Length; i++)
            {
                int size = shipSizes[i];
                bool placed = false;

                while (!placed)
                {
                    Color.SetColor(Color.YELLOW);
                    Console.WriteLine($"\nРазмещение {shipNames[i]} корабля ({size} палубы)");
                    Color.ResetColor();

                    myBoard.Display(true);

                    Console.Write("Введите начальную координату (например, A1): ");
                    string input = Console.ReadLine()?.ToUpper();

                    if (string.IsNullOrEmpty(input) || input.Length < 2)
                    {
                        Color.SetColor(Color.RED);
                        Console.WriteLine("Неверный формат!");
                        Color.ResetColor();
                        continue;
                    }

                    char letter = input[0];
                    if (!int.TryParse(input.Substring(1), out int number))
                    {
                        Color.SetColor(Color.RED);
                        Console.WriteLine("Неверный формат числа!");
                        Color.ResetColor();
                        continue;
                    }

                    int x = letter - 'A';
                    int y = number - 1;

                    if (size > 1)
                    {
                        Console.Write("Выберите ориентацию (H - горизонтально, V - вертикально): ");
                        char orient = char.ToUpper(Console.ReadKey().KeyChar);
                        Console.WriteLine();

                        Orientation orientation = (orient == 'H') ? Orientation.Horizontal : Orientation.Vertical;
                        Ship ship = new Ship(size, new Coordinate(x, y), orientation);

                        if (IsValidShipPlacement(size, new Coordinate(x, y), orientation))
                        {
                            if (myBoard.PlaceShip(ship))
                            {
                                ships.Add(ship);
                                placed = true;
                                Color.SetColor(Color.GREEN);
                                Console.WriteLine("Корабль размещен успешно!");
                                Color.ResetColor();
                            }
                        }
                        else
                        {
                            Color.SetColor(Color.RED);
                            Console.WriteLine("Неверное размещение корабля! Попробуйте снова.");
                            Color.ResetColor();
                        }
                    }
                    else
                    {
                        Ship ship = new Ship(size, new Coordinate(x, y), Orientation.Horizontal);

                        if (IsValidShipPlacement(size, new Coordinate(x, y), Orientation.Horizontal))
                        {
                            if (myBoard.PlaceShip(ship))
                            {
                                ships.Add(ship);
                                placed = true;
                                Color.SetColor(Color.GREEN);
                                Console.WriteLine("Корабль размещен успешно!");
                                Color.ResetColor();
                            }
                        }
                        else
                        {
                            Color.SetColor(Color.RED);
                            Console.WriteLine("Неверное размещение корабля! Попробуйте снова.");
                            Color.ResetColor();
                        }
                    }
                }
            }
        }

        private bool AutomaticPlacement()
        {
            int attempts = 0;
            const int MAX_ATTEMPTS = 5;
            Random random = new Random();

            while (attempts < MAX_ATTEMPTS)
            {
                myBoard.ClearBoard();
                ships.Clear();
                bool success = true;

                foreach (int size in shipSizes)
                {
                    bool placed = false;
                    int placementAttempts = 0;

                    while (!placed && placementAttempts < MAX_SINGLE_PLACEMENT_TRIES)
                    {
                        int x = random.Next(BOARD_SIZE);
                        int y = random.Next(BOARD_SIZE);
                        Orientation orientation = (random.Next(2) == 0) ?
                            Orientation.Horizontal : Orientation.Vertical;

                        Ship ship = new Ship(size, new Coordinate(x, y), orientation);

                        if (myBoard.IsValidPlacement(ship))
                        {
                            if (myBoard.PlaceShip(ship))
                            {
                                ships.Add(ship);
                                placed = true;
                            }
                        }
                        placementAttempts++;
                    }

                    if (!placed)
                    {
                        success = false;
                        break;
                    }
                }

                if (success)
                {
                    Color.SetColor(Color.GREEN);
                    Console.WriteLine("Автоматическая расстановка завершена!");
                    Color.ResetColor();
                    myBoard.Display(true);

                    if (attempts > 0)
                    {
                        Console.WriteLine($"Потребовалось {attempts + 1} попыток для успешной расстановки.");
                    }
                    return true;
                }
                attempts++;
            }

            Color.SetColor(Color.RED);
            Console.WriteLine($"Не удалось автоматически расставить корабли после {MAX_ATTEMPTS} попыток.");
            Color.ResetColor();
            return false;
        }

        private bool IsValidShipPlacement(int size, Coordinate start, Orientation orientation)
        {
            Ship tempShip = new Ship(size, start, orientation);
            return myBoard.IsValidPlacement(tempShip);
        }
    }
}